### ISSUE TITLE
Not Able to send push notification 

### ISSUE THREAD

**Issue by sukritisuku** [noodlewerk/NWPusher#76]
Hi,

I am not able to send push notification any more from Pusher Mac app. It was showing Notification error: APN invalid Token. please give the solution for the issue

**Comment by mroffmix** [comment]
Hi, if you are on ios 13 you can try to format your token using this line of code.
```swift
let deviceToken = credentials.token.map { String(format: "%02x", $0) }.joined()
```

**Comment by sukritisuku** [comment]
hi 
i am not using iOS 13. I am on iOS 12. still not able to get push.

Error = "Notification error: APN invalid Token"

**Comment by dhruvil1514** [comment]
Same issue I am also facing

**Comment by saormart** [comment]
**iOS 13
MacOS Catalina
Xcode 11.3.1**
For me is working fine from Mac OS or iOS on iPhones / iPads!
Great project! Thanks !

**Comment by brunomunizaf** [comment]
on ```didRegisterForRemoteNotificationsWithDeviceToken``` do: 
```
let tokenParts = deviceToken.map { data -> String in
    return String(format: "%02.2hhx", data)
}

let token = tokenParts.joined()
```

**Comment by enasaz9** [comment]
@brunomunizaf Not working with me.
any update?

**Comment by brunomunizaf** [comment]
> @brunomunizaf Not working with me.
> 
> any update?


Nope. Just gave up... alternatively used a script. Been a while - i probably lost it already. Will look into it

**Comment by enasaz9** [comment]
@brunomunizaf thank you so much, please share this script with me if possible, your help is highly appreciated :)

**Comment by brunomunizaf** [comment]
@enasaz9 try this out https://github.com/tsif/pu.sh/blob/master/pu.sh worked for me.

**Comment by pacocd** [comment]
In case `APN invalid Token` appears, make sure that you are using correct environment and proper extraction of token.
1.- `Should use sandbox environment` enabled works for devices using debug mode.
2.- On recent iOS versions, the way to get device string tokens has been updated due to changes in `Data` class `deviceToken.reduce("", { $0 + String(format: "%02.2hhx", $1) })` (I am not sure at all if this changed on iOS 12 or 13)

Hope this can help you :)
